### PR TITLE
Fix `loss_fn` in librispeech workloads

### PR DESCRIPTION
**Pull request by runame**
This PR is a follow-up to #234 and fixes the `loss_fn` in the librispeech workloads to adhere to the new spec. Moreover, I noticed that in the PyTorch workloads the loss was synced via `all_reduce` in `loss_fn`. This is a bug, since it introduces unnecessary communication overhead when using PyTorch DDP (and can maybe even lead to wrong behavior, not sure about that).

### CODE REVIEW

**Review by znado — APPROVED** [review]

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py [hunk 1]
```diff
@@ -93,23 +93,33 @@ def model_fn(
...
+                                       logit_paddings,
+                                       targets,
+                                       target_paddings)
+    # mask_batch is assumed to be shape [batch].
+    if mask_batch is not None:
+      per_example_losses *= mask_batch
+      mask_batch = jnp.logical_and(mask_batch, 1 - target_paddings)
+    else:
+      mask_batch = target_paddings
+    n_valid_examples = jnp.maximum(mask_batch.sum(), 1)
```
Comment (sourabh2k15) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:120]: this should be `n_valid_examples  = jnp.maximum(1 - mask_batch.sum(), 1)`,  given mask_batch is target_paddings if it's None

in librispeech padding 1.0 means padded value

Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:120]: Great catch, fixed!

**Comment by sourabh2k15** [comment]
Hi Runa, I wanted to better understand the pytorch part before we merge this ... are we sure this is a bug and we need to fix it ? 

asking cause Chandramouli has a test comparing previous state of jax librispeech loss function to previous state of torch librispeech loss function 

at the very least we need to run that test and confirm it passes, I'd also request you to run the updated loss functions and make sure they get exactly same values as before

**Comment by sourabh2k15** [comment]
also you could include changes from https://github.com/mlcommons/algorithmic-efficiency/pull/241 

I'll close that PR if those changes are included here , but it's absolutely needed after loss API changes , currently librispeech workloads are broken at head

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 2]
```diff
@@ -266,15 +266,18 @@ def _eval_model_on_split(self,
...
+          decoded=decoded.cpu().numpy(),
+          decoded_paddings=decoded_paddings.cpu().numpy(),
+          targets=targets.cpu().numpy(),
+          target_paddings=target_paddings.cpu().numpy(),
           tokenizer=self.tokenizer)
-      loss = self._loss_fn((targets, target_paddings), (logits, logits_padding))
+      mean_loss, per_example_losses = self.loss_fn(
+          (targets, target_paddings), (logits, logits_padding))
+      summed_loss = per_example_losses.sum()
+      lengths = summed_loss / mean_loss
```
Comment (chandramouli-sastry) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:277]: While this is correct theoretically, we might need to do a round operation to ensure that the `lengths` are whole numbers? In the eval step, we `all_reduce` the lengths and then use it for computing globally normalized ctc loss and am not totally sure if this would distort the loss-values in some way and by how much -- at this point, im okay either way :)

**Comment by runame** [comment]
> The all_reduce in the loss function is not a bug, but i agree that it is a overhead. Essentially, average of normalized-ctc-losses is not equal to the globally normalized ctc-loss. On backward, all the processes will have the same gradients for all parameters and since the DDP takes care of averaging the gradient values, i do not think that this will be a problem?
> 
> I am okay with this change as it was not there to begin with. But, if this is the reason for difference between jax and pytorch, i can put this back in at a later point?

Yeah with bug I mostly meant the additional communication overhead (I only mentioned that I'm not sure if it can lead to bad behavior, because I wasn't sure if there can be issues with backpropping through an `all_reduce` operation).

But I see, so you are saying without globally calculating the loss, DDP will not match the correct global gradient. Then I'm fine with adding the syncing code back in, if you haven't seen any issues with it.

**Comment by runame** [comment]
The purpose of this PR was to fix this issue, so I think it is already included?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 2 above)
Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:277]: Yeah, it doesn't hurt to round here. I'll fix it, thanks!

**Comment by chandramouli-sastry** [comment]
> > The all_reduce in the loss function is not a bug, but i agree that it is a overhead. Essentially, average of normalized-ctc-losses is not equal to the globally normalized ctc-loss. On backward, all the processes will have the same gradients for all parameters and since the DDP takes care of averaging the gradient values, i do not think that this will be a problem?
> > I am okay with this change as it was not there to begin with. But, if this is the reason for difference between jax and pytorch, i can put this back in at a later point?
> 
> Yeah with bug I mostly meant the additional communication overhead (I only mentioned that I'm not sure if it can lead to bad behavior, because I wasn't sure if there can be issues with backpropping through an `all_reduce` operation).
> 
> But I see, so you are saying without globally calculating the loss, DDP will not match the correct global gradient. Then I'm fine with adding the syncing code back in, if you haven't seen any issues with it.

Yes, there is a longer discussion about that here: https://github.com/pytorch/pytorch/issues/58005

I went ahead and used dist.all_reduce, but the alternative is to use nn.all_reduce instead: but in one of the previous runs with dist.all_reduce, it seemed like there was no harm done. I am okay with this being removed as of now -- we can add this back in after carefully reviewing the effects.

**Review by znado — APPROVED** [review]

**Comment by znado** [comment]
I will let @sourabh2k15 comment on if we should sync the CTC loss during training or not. If it is not mathematically equivalent to the single device case, we need to carefully and explicitly document this (in the function comments, in the workload module docstrings, in the README, and other places) because it is a big diff.

**Comment by chandramouli-sastry** [comment]
> > > The all_reduce in the loss function is not a bug, but i agree that it is a overhead. Essentially, average of normalized-ctc-losses is not equal to the globally normalized ctc-loss. On backward, all the processes will have the same gradients for all parameters and since the DDP takes care of averaging the gradient values, i do not think that this will be a problem?
> > > I am okay with this change as it was not there to begin with. But, if this is the reason for difference between jax and pytorch, i can put this back in at a later point?
> > 
> > 
> > Yeah with bug I mostly meant the additional communication overhead (I only mentioned that I'm not sure if it can lead to bad behavior, because I wasn't sure if there can be issues with backpropping through an `all_reduce` operation).
> > But I see, so you are saying without globally calculating the loss, DDP will not match the correct global gradient. Then I'm fine with adding the syncing code back in, if you haven't seen any issues with it.
> 
> Yes, there is a longer discussion about that here: [pytorch/pytorch#58005](https://github.com/pytorch/pytorch/issues/58005)
> 
> I went ahead and used dist.all_reduce, but the alternative is to use nn.all_reduce instead: but in one of the previous runs with dist.all_reduce, it seemed like there was no harm done. I am okay with this being removed as of now -- we can add this back in after carefully reviewing the effects.

From the recent conformer-tests, we have identified that we should use dist.nn.all_reduce to match the jax implementation.

**Comment by znado** [comment]
For speech (and WMT, and any workload with masked training batches), won't we still have an issue with not properly averaging the loss across devices? In the target setting (and dev) algo we take the mean of the per-device losses [here](https://github.com/mlcommons/algorithmic-efficiency/blob/main/reference_algorithms/target_setting_algorithms/jax_submission_base.py#L49), but because each of those is a weighted mean, won't this not be correct?
